crates/polars-ops/src/series/ops/unique.rs

```diff
@@ -25,6 +26,110 @@ where
     out.into_inner()
 }
 
+fn unique_counts_boolean_helper(ca: &BooleanChunked) -> IdxCa {
+    if ca.is_empty() {
+        return IdxCa::new(ca.name(), [] as [IdxSize; 0]);
+    }
+
+    let ca = ca.rechunk();
+    let arr = ca.downcast_iter().next().unwrap();
+
+    let (n_true, n_null);
+    if let Some(validity) = arr.validity() {
+        n_null = validity.unset_bits();
+        if n_null < arr.len() {
+            n_true = (arr.values() & validity).set_bits();
+        } else {
+            n_true = 0;
+        }
+    } else {
+        n_null = 0;
+        n_true = arr.values().set_bits();
+    }
+    let n_false = arr.len() - n_true - n_null;
+
+    if n_true == 0 && n_false == 0 {
+        return IdxCa::new(ca.name(), [n_null as IdxSize]);
+    }
+    if n_true == 0 && n_null == 0 {
+        return IdxCa::new(ca.name(), [n_false as IdxSize]);
+    }
+    if n_false == 0 && n_null == 0 {
+        return IdxCa::new(ca.name(), [n_true as IdxSize]);
+    }
+
+    if n_true == 0 {
+        match arr.is_null(0) {
+            true => return IdxCa::new(ca.name(), [n_null as IdxSize, n_false as IdxSize]),
+            false => return IdxCa::new(ca.name(), [n_false as IdxSize, n_null as IdxSize]),
+        }
+    } else if n_false == 0 {
+        match arr.is_null(0) {
+            true => return IdxCa::new(ca.name(), [n_null as IdxSize, n_true as IdxSize]),
+            false => return IdxCa::new(ca.name(), [n_true as IdxSize, n_null as IdxSize]),
+        }
+    } else if n_null == 0 {
+        match arr.value(0) {
+            true => return IdxCa::new(ca.name(), [n_true as IdxSize, n_false as IdxSize]),
+            false => return IdxCa::new(ca.name(), [n_false as IdxSize, n_true as IdxSize]),
+        }
+    }
+
```

Comment:
Yes, probably a bit easier to follow. However, the code you propose complains about the assignment to the out variable: `'if' and 'else' have incompatible types
   expected an array with a fixed size of 1 element, found one with 2 elements [E0308]`

I guess we'd have to do something like below, unless there are any other ways you know of to handle that the returned array can differ in size. Any preference?

```rust
    if n_true == 0 {
        let out = if n_false == 0 {
            vec![n_null as IdxSize]
        } else if n_null == 0 {
            vec![n_false as IdxSize]
        } else if arr.is_null(0) {
            vec![n_null as IdxSize, n_false as IdxSize]
        } else {
            vec![n_false as IdxSize, n_null as IdxSize]
        };
        return IdxCa::new(ca.name(), out);
    }
```

or 

```rust
    if n_true == 0 {
        if n_false == 0 {
            return IdxCa::new(ca.name(), [n_null as IdxSize]);
        } else if n_null == 0 {
            return IdxCa::new(ca.name(), [n_false as IdxSize]);
        } else if arr.is_null(0) {
            return IdxCa::new(ca.name(), [n_null as IdxSize, n_false as IdxSize]);
        } else {
            return IdxCa::new(ca.name(), [n_false as IdxSize, n_null as IdxSize]);
        };
    }
```

Reply:
Ahh, indeed, sorry; I just wrote that in the browser. It may look a bit cleaner to cast your initial `n_null`, `n_true`, and `n_false` as `IdxSize` so you don't have to litter everything with `as IdxSize`. Will this work?

```rust
    if n_true == 0 {
        return IdxCa::new(ca.name(), if n_false == 0 {
                [n_null]
            } else if n_null == 0 {
                [n_false]
            } else if arr.is_null(0) {
                [n_null, n_false]
            } else {
                [n_false, n_null]
            }
        );
    }
```

Reply:
It still complains about the different return types in the if/else, unfortunately. The initial cast to `IdxSize` is a good idea though!

Reply:
Does it work if you ref the slices? e.g. `&[n_null]`? Sorry I'm at work so I can't test these myself. I'm not super familiar with rust so these things aren't ingrained for me :/

Reply:
Alright. I'd just go with e.g. `return IdxCa::new(ca.name(), [n_null]);` in each case then.

Reply:
Just realized your proposal also misses the case when there are only true/false and no nulls, so the two implementations do not differ that much anymore and are about equally many lines. Here's a comparison:

```rust
    // original (with `IdxSize` removed)
    if n_true == 0 && n_false == 0 {
        return IdxCa::new(ca.name(), [n_null]);
    }
    if n_true == 0 && n_null == 0 {
        return IdxCa::new(ca.name(), [n_false]);
    }
    if n_false == 0 && n_null == 0 {
        return IdxCa::new(ca.name(), [n_true]);
    }

    if n_true == 0 {
        match arr.is_null(0) {
            true => return IdxCa::new(ca.name(), [n_null, n_false]),
            false => return IdxCa::new(ca.name(), [n_false, n_null]),
        }
    } else if n_false == 0 {
        match arr.is_null(0) {
            true => return IdxCa::new(ca.name(), [n_null, n_true]),
            false => return IdxCa::new(ca.name(), [n_true, n_null]),
        }
    } else if n_null == 0 {
        match arr.value(0) {
            true => return IdxCa::new(ca.name(), [n_true, n_false]),
            false => return IdxCa::new(ca.name(), [n_false, n_true]),
        }
    }
```

VS

```rust
    // proposal
    if n_true == 0 {
        if n_false == 0 {
            return IdxCa::new(ca.name(), [n_null]);
        } else if n_null == 0 {
            return IdxCa::new(ca.name(), [n_false]);
        } else if arr.is_null(0) {
            return IdxCa::new(ca.name(), [n_null, n_false]);
        } else {
            return IdxCa::new(ca.name(), [n_false, n_null]);
        };
    }

    if n_false == 0 {
        if n_null == 0 {
            return IdxCa::new(ca.name(), [n_true]);
        } else if arr.is_null(0) {
            return IdxCa::new(ca.name(), [n_null, n_true]);
        } else {
            return IdxCa::new(ca.name(), [n_true, n_null]);
        };
    }

    if n_null == 0 {
        match arr.value(0) {
            true => return IdxCa::new(ca.name(), [n_true, n_false]),
            false => return IdxCa::new(ca.name(), [n_false, n_true]),
        }
    }
```

Tbh I think I prefer the original implementation due to the avoidance of nested `if` statements and the `else if`'s that can be hard to track, but that might just be me and because I wrote the code myself. Open to changing if you guys find the other version more readable.

Reply:
Yeah, it's not a huge deal and I doubt will make any difference in time since these are just simple if statements on scalar logicals.